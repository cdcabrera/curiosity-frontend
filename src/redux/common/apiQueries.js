import { rhsmApiTypes } from '../../types/rhsmApiTypes';
import { reduxHelpers } from './reduxHelpers';

/**
 * Parse a query object against a schema for specific RHSM endpoints.
 *
 * @param {object} query
 * @returns {{query: {}, guestsQuery: {}, graphQuery: {}, inventoryQuery: {}, toolbarQuery: {}}}
 */
const parseRhsmQuery = (query = {}) => {
  const graphQuery = reduxHelpers.setApiQuery(query, rhsmApiTypes.RHSM_API_QUERY_SET_REPORT_CAPACITY_TYPES);
  const guestsQuery = reduxHelpers.setApiQuery(query, rhsmApiTypes.RHSM_API_QUERY_SET_INVENTORY_GUESTS_TYPES);
  const inventoryQuery = reduxHelpers.setApiQuery(query, rhsmApiTypes.RHSM_API_QUERY_SET_INVENTORY_TYPES);
  const pagingQuery = reduxHelpers.setApiQuery(query, rhsmApiTypes.RHSM_API_QUERY_SET_PAGING_TYPES);

  return {
    query,
    graphQuery,
    guestsQuery,
    inventoryQuery,
    pagingQuery,
    toolbarQuery: query
  };
};

const apiQueries = {
  parseRhsmQuery
};

export { apiQueries as default, apiQueries };
