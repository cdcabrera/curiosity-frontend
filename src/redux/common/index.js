import { reduxHelpers } from './reduxHelpers';

/**
 * @memberof Redux State
 * @module Helpers
 * @property {module} ReduxHelpers
 */

export { reduxHelpers as default, reduxHelpers };
