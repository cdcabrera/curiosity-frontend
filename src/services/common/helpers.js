import _camelCase from 'lodash/camelCase';
import _isPlainObject from 'lodash/isPlainObject';

const camelCase = obj => {
  if (Array.isArray(obj)) {
    return obj.map(camelCase);
  }

  if (_isPlainObject(obj)) {
    const updatedObj = {};

    Object.entries(obj).forEach(([key, val]) => {
      updatedObj[_camelCase(key)] = camelCase(val);
    });

    return updatedObj;
  }

  return obj;
};

const schemaResponse = ({ convert = true, id = null, response, schema }) => {
  const { value, error = { details: [] } } = schema.validate(response, { convert });

  if (error.details.length) {
    throw new Error(
      `Schema validation for ${id || '...'}: ${error.details
        .map(({ message, type }) => `${message}:${type}`)
        .join(', ')}`
    );
  }

  return camelCase(value);
};

const serviceHelpers = {
  camelCase,
  schemaResponse
};

export { serviceHelpers as default, camelCase, schemaResponse };
