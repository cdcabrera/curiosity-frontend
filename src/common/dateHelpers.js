import moment from 'moment/moment';
import { helpers } from './helpers';
import { rhelApiTypes } from '../types/rhelApiTypes';

const getCurrentDate = () =>
  (helpers.TEST_MODE && '20190720') || (helpers.DEV_MODE && process.env.REACT_APP_DEBUG_DEFAULT_DATETIME) || new Date();

const setRangedDateTime = ({ date, subtract, measurement }) => ({
  start: moment
    .utc(date)
    .startOf('day')
    .subtract(subtract, measurement)
    .toDate(),
  end: moment
    .utc(date)
    .endOf('day')
    .toDate()
});

const defaultDateTime = setRangedDateTime({ date: getCurrentDate(), subtract: 30, measurement: 'days' });
const weeklyDateTime = setRangedDateTime({ date: getCurrentDate(), subtract: 12, measurement: 'weeks' });
const monthlyDateTime = setRangedDateTime({ date: getCurrentDate(), subtract: 12, measurement: 'months' });
const quarterlyDateTime = setRangedDateTime({ date: getCurrentDate(), subtract: 4, measurement: 'years' });

/**
 * Return a range of time based on granularity.
 *
 * @param {string} granularity
 * @returns {{endDate: Date, startDate: Date}}
 */
const GRANULARITY_TYPES = rhelApiTypes.RHSM_API_QUERY_GRANULARITY_TYPES;

const getRangedDateTime = granularity => {
  switch (granularity) {
    case GRANULARITY_TYPES.WEEKLY:
      return { startDate: weeklyDateTime.start, endDate: weeklyDateTime.end };
    case GRANULARITY_TYPES.MONTHLY:
      return { startDate: monthlyDateTime.start, endDate: monthlyDateTime.end };
    case GRANULARITY_TYPES.QUARTERLY:
      return { startDate: quarterlyDateTime.start, endDate: quarterlyDateTime.end };
    case GRANULARITY_TYPES.DAILY:
    default:
      return { startDate: defaultDateTime.start, endDate: defaultDateTime.end };
  }
};

const dateHelpers = {
  getCurrentDate,
  getRangedDateTime,
  setRangedDateTime,
  defaultDateTime,
  monthlyDateTime,
  quarterlyDateTime,
  weeklyDateTime
};

export {
  dateHelpers as default,
  getCurrentDate,
  getRangedDateTime,
  setRangedDateTime,
  dateHelpers,
  defaultDateTime,
  monthlyDateTime,
  quarterlyDateTime,
  weeklyDateTime
};
