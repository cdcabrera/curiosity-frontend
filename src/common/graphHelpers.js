import moment from 'moment';
import { rhelApiTypes } from '../types/rhelApiTypes';
import { helpers } from './helpers';
import { translate } from '../components/i18n/i18n';

const GRANULARITY_TYPES = rhelApiTypes.RHSM_API_QUERY_GRANULARITY_TYPES;

/**
 * Chart Date Format (used in axis and tooltips)
 */
const chartDateDayFormatShort = 'MMM D';
const chartDateDayFormat = 'MMM D YYYY';

const chartDateMonthFormatShort = 'MMM';
const chartDateMonthFormat = 'MMM YYYY';

/**
 * Generate a fallback graph with zeroed data
 *
 * @param {string} startDate
 * @param {string} endDate
 * @returns {Array}
 */
/*
const zeroedUsageData = (startDate, endDate) => {
  const zeroedArray = [];
  const endDateStartDateDiff = moment(endDate).diff(startDate, 'days');

  for (let i = 0; i <= endDateStartDateDiff; i++) {
    const clonedStartDate = moment.utc(startDate);
    zeroedArray.push({
      x: clonedStartDate.add(i, 'days').format(chartDateFormat),
      y: 0
    });
  }

  return zeroedArray;
};
*/

/**
 * Returns y axis range, aka chart domain, setting for given inputs.
 *
 * the y axis max should be rounded to the nearest 10 if less than 50,
 * otherwise round to the nearest power of 10
 *
 * the y axis returns large enough number that zeroed bars dont show
 *
 * @param {number} chartDataLength Number of chart data points
 * @param {boolean} maxY The max y-value
 * @returns {Object}
 */
const getChartDomain = ({ chartDataLength, maxY = 0 }) => {
  // const chartDomain = { x: [0, chartDataLength === 0 ? 31 : chartDataLength + 1] };
  // const chartDomain = { x: [0, chartDataLength === 0 ? 31 : chartDataLength / 3] };
  const chartDomain = {};

  if (chartDataLength === 0 || maxY < 50) {
    chartDomain.y = [0, Math.ceil((maxY + 1) / 10) * 10];
  } else {
    chartDomain.y = [0, Math.pow(10, maxY.toString().length)];
  }

  return chartDomain;
};

/**
 * Return a time allotment based on granularity
 *
 * @param {string} granularity enum based on rhelApiTypes.RHSM_API_QUERY_GRANULARITY_TYPES
 * @returns {string}
 */
const getGranularityType = granularity => {
  switch (granularity) {
    case GRANULARITY_TYPES.DAILY:
      return 'days';
    case GRANULARITY_TYPES.WEEKLY:
      return 'weeks';
    case GRANULARITY_TYPES.MONTHLY:
    case GRANULARITY_TYPES.QUARTERLY:
    default:
      return 'months';
  }
};

const getGraphLabels = granularity => {
  const labels = {
    label: translate('curiosity-graph.tooltipLabel')
  };

  switch (granularity) {
    case GRANULARITY_TYPES.WEEKLY:
      labels.previousLabel = translate('curiosity-graph.tooltipPreviousLabelWeekly');
      break;
    case GRANULARITY_TYPES.MONTHLY:
      labels.previousLabel = translate('curiosity-graph.tooltipPreviousLabelMonthly');
      break;
    case GRANULARITY_TYPES.QUARTERLY:
      labels.previousLabel = translate('curiosity-graph.tooltipPreviousLabelQuarterly');
      break;
    case GRANULARITY_TYPES.DAILY:
    default:
      labels.previousLabel = translate('curiosity-graph.tooltipPreviousLabelDaily');
      break;
  }

  return labels;
};

/**
 * Apply label formatting
 *
 * @param {number} data
 * @param {number} previousData
 * @param {string} formattedDate
 * @param {string} label
 * @param {string} previousLabel
 * @returns {string}
 */
const getLabel = ({ data, previousData, formattedDate, granularity }) => {
  const { label, previousLabel } = getGraphLabels(granularity);
  const previousCount = data - previousData;
  const updatedLabel = `${data} ${label} ${formattedDate}`;

  if (!previousData || previousCount === 0) {
    return updatedLabel;
  }

  return `${updatedLabel}\n ${previousCount > -1 ? '+' : ''}${previousCount} ${previousLabel}`;
};

/**
 * Returns x axis ticks/intervals array for the xAxisTickInterval
 *
 * @param {Array} chartData
 * @param {number} xAxisTickInterval
 * @param {(days|weeks|months)} granularity
 * @returns {Array}
 */
const getTickValues = ({ chartData, xAxisTickInterval = 5, granularity }) => {
  // const tickInterval = granularityType === 'weeks' || granularityType === 'months' ? 1 : xAxisTickInterval;
  // const tickInterval = chartData.length < 15 ? 1 : xAxisTickInterval;
  let tickInterval = xAxisTickInterval;

  switch (granularity) {
    case GRANULARITY_TYPES.DAILY:
      tickInterval = 5;
      break;
    case GRANULARITY_TYPES.QUARTERLY:
      // tickInterval = 5;
      tickInterval = 1;
      break;
    case GRANULARITY_TYPES.WEEKLY:
    case GRANULARITY_TYPES.MONTHLY:
    default:
      tickInterval = 1;
      break;
  }

  // debugger;
  // const ticks = chartData.reduce((acc, current, index) => ((chartData.length < 15) || (index % tickInterval === 0) ? acc.concat(current.x) : acc), []);
  const ticks = chartData.reduce(
    (acc, current, index) => (index % tickInterval === 0 ? acc.concat(current.x) : acc),
    []
  );
  // return chartData.map(value => value.x);
  // debugger;
  console.log('TICKS - - - - -- - ------>', granularity, ticks);

  return ticks;
};

/**
 * Fills missing dates with zeroed y-values, updates labels
 * walk the date range and fill missing values, all the while updating previous labels
 *
 * @param {string} startDate
 * @param {string} endDate
 * @param {object} values pre-filled key-value object
 * @param {(days|weeks|months)} granularityType
 * @param {string} label i18n specific label
 * @param {string} previousLabel i18n specific previousLabel
 * @returns {Array}
 */
const fillMissingValuesOLD = ({ startDate, endDate, values, granularityType, label, previousLabel }) => {
  const endDateStartDateDiff = moment(endDate).diff(startDate, granularityType);
  const chartData = [];
  let previousYear = null;

  for (let i = 0; i <= endDateStartDateDiff; i++) {
    const momentDate = moment.utc(startDate).add(i, granularityType);
    const stringDate = momentDate.toISOString();
    const year = parseInt(momentDate.year(), 10);
    const isNewYear = previousYear !== null && year !== previousYear;
    const formattedDate = momentDate.format((isNewYear && chartDateDayFormat) || chartDateDayFormatShort);

    const retData = {
      data: 0,
      previousData: i > 0 ? chartData[i - 1].y : null,
      formattedDate,
      label,
      previousLabel
    };

    if (values[stringDate]) {
      retData.data = values[stringDate].y;

      chartData.push({
        x: values[stringDate].x,
        y: values[stringDate].y,
        label: getLabel(retData)
      });
    } else {
      chartData.push({
        x: formattedDate,
        y: 0,
        label: getLabel(retData)
      });
    }

    previousYear = year;
  }

  // debugger;

  /*
  for (let i = 0; i <= endDateStartDateDiff; i++) {
    const formattedDate = moment
      .utc(startDate)
      .add(i, granularityType)
      .format(chartDateFormat);

    const updatedLabel = getLabel({
      data: values[formattedDate] ? values[formattedDate].y : 0,
      previousData: i > 0 ? chartData[i - 1].y : null,
      formattedDate,
      label,
      previousLabel
    });

    if (values[formattedDate]) {
      chartData.push({
        x: values[formattedDate].x,
        y: values[formattedDate].y,
        label: updatedLabel
      });
    } else {
      chartData.push({
        x: formattedDate,
        y: 0,
        label: updatedLabel
      });
    }
  }
  */

  return chartData;
};

const getChartData = ({ data, endDate, granularity, startDate }) => {
  const granularityType = getGranularityType(granularity);
  const endDateStartDateDiff = moment(endDate).diff(startDate, granularityType);
  const chartData = [];

  let previousData = null;
  let previousYear = null;

  console.log('REWRITE 1A - - - - -- - - ----->', data, endDateStartDateDiff);

  for (let i = 0; i <= endDateStartDateDiff; i++) {
    if (GRANULARITY_TYPES.QUARTERLY === granularity && i % 4 !== 0) {
      continue;
    }

    const momentDate = moment.utc(startDate).add(i, granularityType);
    const stringDate = momentDate.toISOString();
    const year = parseInt(momentDate.year(), 10);

    const isNewYear = year !== previousYear;
    let formattedDate;

    if (granularityType === 'months') {
      formattedDate = isNewYear
        ? momentDate.format(chartDateMonthFormat)
        : momentDate.format(chartDateMonthFormatShort);
    } else {
      formattedDate = isNewYear ? momentDate.format(chartDateDayFormat) : momentDate.format(chartDateDayFormatShort);
    }

    console.log('REWRITE 1B - - - - -- - - ----->', year, isNewYear, stringDate);

    const labelData = {
      data: data[stringDate] || 0,
      // previousData: i > 0 ? chartData[i - 1].y : null,
      previousData,
      formattedDate,
      // label,
      // previousLabel,
      granularity
    };

    chartData.push({
      x: formattedDate,
      y: data[stringDate] || 0,
      label: getLabel(labelData)
    });

    previousData = data[stringDate];
    previousYear = year;
  }

  return chartData;
};

/**
 * Convert graph data to usable format
 * convert json usage report from this format:
 *  {cores: 56, date: "2019-06-01T00:00:00Z", instance_count: 28}
 * to this format:
 *  { x: 'Jun 1', y: 56, label: '56 Sockets on Jun 1 \r\n +5 from previous day' }
 *
 * @param {Array} data
 * @param {date} startDate
 * @param {date} endDate
 * @param {string} granularity enum of rhelApiTypes.RHSM_API_QUERY_GRANULARITY_TYPES
 * @param {string} label
 * @param {string} previousLabel
 * @param {number} xAxisTickInterval
 * @returns {Object} Object array result converted { chartData: {...} chartDomain {...} }
 */
const convertGraphUsageData = ({ data, startDate, endDate, granularity, label, previousLabel, xAxisTickInterval }) => {
  const formattedData = {};
  let maxY = 0;

  (data || []).forEach(value => {
    const stringDate = moment.utc(value[rhelApiTypes.RHSM_API_RESPONSE_PRODUCTS_DATA_DATE]).toISOString();

    formattedData[stringDate] = Number.parseInt(value[rhelApiTypes.RHSM_API_RESPONSE_PRODUCTS_DATA_SOCKETS], 10);
    maxY = formattedData[stringDate] > maxY ? formattedData[stringDate] : maxY;
  });

  console.log('REWRITE 1 - - - - -- - - ----->', formattedData);

  const chartData = getChartData({ data: formattedData, endDate, granularity, startDate, label, previousLabel });

  console.log('REWRITE 2 - - - - -- - - ----->', chartData);

  const chartDomain = getChartDomain({ chartDataLength: chartData.length, maxY });
  const tickValues = getTickValues({ chartData, xAxisTickInterval, granularity });

  console.log('REWRITE 3 - - - - -- - - ----->', chartData, chartDomain, tickValues);

  return {
    chartData,
    chartDomain,
    tickValues
  };
};

/*
const convertGraphUsageDataOLD = ({ data, startDate, endDate, granularity, label, previousLabel, xAxisTickInterval }) => {
  const granularityType = getGranularityType(granularity);
  const values = {};
  let chartData = [];
  let chartDomain = {};
  let maxY = 0;

  try {
    data.forEach(value => {
      // const formattedDate = moment
      // .utc(value[rhelApiTypes.RHSM_API_RESPONSE_PRODUCTS_DATA_DATE])
      //  .format(chartDateDayFormatShort);

      const momentDate = moment.utc(value[rhelApiTypes.RHSM_API_RESPONSE_PRODUCTS_DATA_DATE]);
      const stringDate = momentDate.toISOString();
      // const year = parseInt(momentDate.year(), 10);
      // const isNewYear = previousYear !== null && year !== previousYear;
      // const formattedDate = momentDate.format((isNewYear && chartDateDayFormat) || chartDateDayFormatShort);

      const entry = {
        x: formattedDate,
        y: Number.parseInt(value[rhelApiTypes.RHSM_API_RESPONSE_PRODUCTS_DATA_SOCKETS], 10)
      };

      maxY = entry.y > maxY ? entry.y : maxY;
      values[entry.x] = entry;
      // chartData.push(entry);
    });
  } catch (e) {
    if (!helpers.TEST_MODE) {
      console.warn(`Malformed API response ${e.message}`);
    }
  }

  chartData = fillMissingValues({ startDate, endDate, values, granularityType, label, previousLabel });
  chartDomain = getChartDomain({ chartDataLength: chartData.length, maxY });
  const tickValues = getTickValues({ chartData, xAxisTickInterval, granularity });

  return {
    chartData,
    chartDomain,
    tickValues
  };

  /*
  const granularityType = getGranularityType(granularity);
  let chartData = [];
  let chartDomain = {};

  try {
    const values = {};
    let maxY = 0;

    for (let i = 0; i < data.length; i++) {
      const formattedDate = moment
        .utc(data[i][rhelApiTypes.RHSM_API_RESPONSE_PRODUCTS_DATA_DATE])
        .format(chartDateFormat);

      values[formattedDate] = {
        x: formattedDate,
        y: data[i][rhelApiTypes.RHSM_API_RESPONSE_PRODUCTS_DATA_SOCKETS]
      };

      maxY = values[formattedDate].y > maxY ? values[formattedDate].y : maxY;
    }

    if (data.length) {
      chartData = fillMissingValues({ startDate, endDate, values, granularityType, label, previousLabel });
    }

    chartDomain = getChartDomain({ entries: chartData.length, maxY });
  } catch (e) {
    if (!helpers.TEST_MODE) {
      console.warn(`Malformed API response ${e.message}`);
    }
  }

  if (!chartData.length) {
    // chartData = zeroedUsageData(startDate, endDate);
    chartData = fillMissingValues({ startDate, endDate, granularityType });
    chartDomain = getChartDomain({ chartData, entries: 0 });
  }

  const tickValues = getTickValues({ chartData, xAxisTickInterval });

  return { chartData, chartDomain, tickValues };
  * /
};
*/

const graphHelpers = {
  // chartDateFormat,
  convertGraphUsageData,
  getChartDomain,
  getGranularityType,
  getTickValues
  // zeroedUsageData
};

export {
  graphHelpers as default,
  graphHelpers,
  // chartDateFormat,
  convertGraphUsageData,
  getChartDomain,
  getGranularityType,
  getTickValues
  // zeroedUsageData
};
