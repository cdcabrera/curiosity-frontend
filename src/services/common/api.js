import axios, { CancelToken } from 'axios';
import LruCache from 'lru-cache';
import { serviceHelpers } from './helpers';

/**
 * Cache Axios service call cancel tokens.
 *
 * @private
 * @type {object}
 */
const cancelTokens = {};

/**
 * Cache Axios service call responses.
 *
 * @type {object}
 */
const responseCache = new LruCache({
  maxAge: Number.parseInt(process.env.REACT_APP_AJAX_CACHE, 10) || 30000,
  max: 100,
  updateAgeOnGet: true
});

/**
 * Set Axios configuration, which includes response schema validation and caching.
 * Call platform "getUser" auth method, and apply service config. Service configuration
 * includes the ability to cancel all and specific calls, cache and normalize a response
 * based on both a provided schema and a successful API response. The cache will refresh
 * its timeout on continuous calls. To reset it a user will either need to refresh the
 * page or wait the "maxAge".
 *
 * @param {object} config
 * @param {boolean} config.authenticate
 * @param {string} config.url
 * @param {object} config.params
 * @param {boolean} config.cache
 * @param {boolean} config.cancel
 * @param {string} config.cancelId
 * @param {Array} config.schema
 * @param {Array} config.transform
 * @returns {Promise<*>}
 */
const serviceCall = async (config = {}) => {
  const updatedConfig = {
    ...config,
    cache: undefined,
    cacheResponse: config.cache
  };

  const responseTransformers = [];
  const cancelledMessage = 'cancelled request';
  const axiosInstance = axios.create();
  const sortedParams =
    (updatedConfig.params && Object.entries(updatedConfig.params).sort(([a], [b]) => a.localeCompare(b))) || [];
  const cacheId = `${updatedConfig.url || ''}-${JSON.stringify(sortedParams)}`;

  if (updatedConfig.cancel === true) {
    const cancelTokensId = `${updatedConfig.cancelId || ''}-${updatedConfig.url}`;

    if (cancelTokens[cancelTokensId]) {
      cancelTokens[cancelTokensId].cancel(cancelledMessage);
    }

    cancelTokens[cancelTokensId] = CancelToken.source();
    updatedConfig.cancelToken = cancelTokens[cancelTokensId].token;

    delete updatedConfig.cancel;
  }

  if (updatedConfig.cacheResponse === true) {
    const cachedResponse = responseCache.get(cacheId);

    if (cachedResponse) {
      updatedConfig.adapter = adapterConfig =>
        Promise.resolve({
          ...cachedResponse,
          status: 304,
          statusText: 'Not Modified',
          config: adapterConfig
        });

      return axiosInstance(updatedConfig);
    }
  }

  if (updatedConfig.schema) {
    responseTransformers.push(updatedConfig.schema);
  }

  if (updatedConfig.transform) {
    responseTransformers.push(updatedConfig.transform);
  }

  responseTransformers.forEach(([successTransform, errorTransform]) => {
    const transformers = [undefined, response => Promise.reject(response)];

    if (successTransform) {
      transformers[0] = response => {
        const updatedResponse = { ...response };
        const { data, error: normalizeError } = serviceHelpers.passDataToCallback(
          updatedResponse.data,
          successTransform
        );

        if (!normalizeError) {
          updatedResponse.data = data;
        }

        return updatedResponse;
      };
    }

    if (errorTransform) {
      transformers[1] = response => {
        const updatedResponse = { ...response };

        if (updatedResponse?.message === cancelledMessage) {
          return Promise.reject(updatedResponse);
        }

        const { data, error: normalizeError } = serviceHelpers.passDataToCallback(
          updatedResponse?.response?.data,
          errorTransform
        );

        if (!normalizeError) {
          updatedResponse.response = { ...updatedResponse.response, data };
        }

        return Promise.reject(updatedResponse);
      };
    }

    axiosInstance.interceptors.response.use(...transformers);
  });

  if (updatedConfig.cacheResponse === true) {
    axiosInstance.interceptors.response.use(
      response => {
        const updatedResponse = { ...response };
        responseCache.set(cacheId, updatedResponse);
        return updatedResponse;
      },
      response => Promise.reject(response)
    );
  }

  if (typeof updatedConfig.url === 'function') {
    const emulateCallback = updatedConfig.url;
    updatedConfig.url = '/';

    let message = `success, emulated`;
    let emulatedResponse;
    let isSuccess = true;

    try {
      emulatedResponse = await emulateCallback();
    } catch (e) {
      isSuccess = false;
      message = e.message;
    }

    if (isSuccess) {
      updatedConfig.adapter = adapterConfig =>
        Promise.resolve({
          data: emulatedResponse,
          status: 200,
          statusText: message,
          config: adapterConfig
        });
    } else {
      updatedConfig.adapter = adapterConfig =>
        Promise.reject({ // eslint-disable-line
          ...new Error(message),
          message,
          status: 418,
          config: adapterConfig
        });
    }
  }

  console.log('API >>>>>>>', updatedConfig);

  return axiosInstance(updatedConfig);
};

const api = { serviceCall };

export { api as default, api, serviceCall };
