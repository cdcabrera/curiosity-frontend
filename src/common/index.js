import { helpers } from './helpers';
import { dateHelpers } from './dateHelpers';
import { downloadHelpers } from './downloadHelpers';

/**
 * @namespace Helpers
 * @property {module} General
 * @property {module} Dates
 * @property {module} Downloads
 */
export { helpers as default, helpers, dateHelpers, downloadHelpers };
