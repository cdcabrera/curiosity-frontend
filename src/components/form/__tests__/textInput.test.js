import React from 'react';
import { mount, shallow } from 'enzyme';
import TextInput from '../textInput';
import { helpers } from '../../../common';

describe('TextInput Component', () => {
  it('should render a basic component', () => {
    const props = {};

    const component = shallow(<TextInput {...props} />);
    expect(component.render()).toMatchSnapshot('basic component');
  });

  it('should handle readOnly, disabled', () => {
    const props = {
      readOnly: true
    };

    const component = mount(<TextInput {...props} />);
    expect(component.render()).toMatchSnapshot('readOnly');

    component.setProps({
      readOnly: false,
      disabled: true
    });

    expect(component.render()).toMatchSnapshot('disabled');

    component.setProps({
      readOnly: false,
      disabled: false
    });

    expect(component.render()).toMatchSnapshot('active');
  });

  it('should return an emulated onChange event', done => {
    const props = {
      value: 'lorem ipsum'
    };

    props.onChange = event => {
      expect(event).toMatchSnapshot('emulated event');
      done();
    };

    const component = mount(<TextInput {...props} />);
    component.find('input').simulate('change');
  });

  it('should return an emulated onClear event on escape', done => {
    const props = {
      value: 'lorem ipsum',
      type: 'search'
    };

    props.onClear = event => {
      expect(event).toMatchSnapshot('emulated event');
      done();
    };

    const component = shallow(<TextInput {...props} />);
    component.instance().onKeyUp({ keyCode: 27, currentTarget: { value: '' }, persist: helpers.noop });
  });

  it('should return an emulated onClear event on search clear', done => {
    const props = {
      value: 'lorem ipsum',
      type: 'search'
    };

    props.onClear = event => {
      expect(event).toMatchSnapshot('emulated event');
      done();
    };

    const component = shallow(<TextInput {...props} />);
    const mockEvent = { currentTarget: { value: 'lorem ipsum' }, persist: helpers.noop };
    component.instance().onMouseUp(mockEvent);
    mockEvent.currentTarget.value = '';
  });
});
