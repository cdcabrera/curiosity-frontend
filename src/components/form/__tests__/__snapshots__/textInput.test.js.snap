// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component should handle readOnly, disabled: active 1`] = `
<input
  aria-invalid="false"
  class="pf-c-form-control curiosity-text-input "
  id="generatedid-"
  type="text"
  value=""
/>
`;

exports[`TextInput Component should handle readOnly, disabled: disabled 1`] = `
<input
  aria-invalid="false"
  class="pf-c-form-control curiosity-text-input "
  disabled=""
  id="generatedid-"
  type="text"
  value=""
/>
`;

exports[`TextInput Component should handle readOnly, disabled: readOnly 1`] = `
<input
  aria-invalid="false"
  class="pf-c-form-control curiosity-text-input "
  id="generatedid-"
  readonly=""
  type="text"
  value=""
/>
`;

exports[`TextInput Component should render a basic component: basic component 1`] = `
<input
  aria-invalid="false"
  class="pf-c-form-control curiosity-text-input "
  id="generatedid-"
  type="text"
  value=""
/>
`;

exports[`TextInput Component should return an emulated onChange event: emulated event 1`] = `
Object {
  "currentTarget": <input
    aria-invalid="false"
    class="pf-c-form-control curiosity-text-input "
    id="generatedid-"
    type="text"
    value="lorem ipsum"
  />,
  "id": "",
  "name": "",
  "persist": [Function],
  "target": <input
    aria-invalid="false"
    class="pf-c-form-control curiosity-text-input "
    id="generatedid-"
    type="text"
    value="lorem ipsum"
  />,
  "value": "lorem ipsum",
}
`;

exports[`TextInput Component should return an emulated onClear event on escape: emulated event 1`] = `
Object {
  "currentTarget": Object {
    "value": "",
  },
  "id": undefined,
  "name": undefined,
  "persist": [Function],
  "target": undefined,
  "value": "",
}
`;

exports[`TextInput Component should return an emulated onClear event on search clear: emulated event 1`] = `
Object {
  "currentTarget": Object {
    "value": "",
  },
  "id": undefined,
  "name": undefined,
  "persist": [Function],
  "target": undefined,
  "value": "",
}
`;
